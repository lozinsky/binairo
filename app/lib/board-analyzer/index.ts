export * from './analyze-board-by-kind';
export * from './analyze-board';
