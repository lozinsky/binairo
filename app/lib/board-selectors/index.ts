export * from './select-filled-imbalanced-line';
export * from './select-line-that-can-be-equal-to-another';
export * from './select-line-that-includes-another';
export * from './select-line-that-is-equal-to-others';
export * from './select-middle-cell-between-two-identical';
export * from './select-next-cell-after-two-identical';
export * from './select-three-or-more-identical-sequential-cells';
export * from './select-unfilled-balanced-line';
