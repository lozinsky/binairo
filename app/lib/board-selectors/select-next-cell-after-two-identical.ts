import type { Board } from '~/lib/board';

import { selectNextCellAfterTwoIdentical as select } from './board-line-selectors';

export function selectNextCellAfterTwoIdentical(target: Board) {
  for (const board of target.toDoubleRotation()) {
    for (const line of board) {
      const payload = select(line);

      if (payload === undefined) {
        continue;
      }

      return { cell: payload.cell, orientation: board.orientation };
    }
  }

  return;
}
