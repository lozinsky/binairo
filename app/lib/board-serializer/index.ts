export * from './board-serializer-collection';
