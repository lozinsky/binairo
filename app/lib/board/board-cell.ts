export type BoardCellKind = 'fixed' | 'regular';

export type BoardCellKindRaw = 0 | 1;

export type BoardCellState = 'B' | 'E' | 'R';

export type BoardCellStateRaw = 0 | 1 | 2;

export function isBoardCellKindRaw(value: unknown): value is BoardCellKindRaw {
  return typeof value === 'number' && value in BOARD_CELL_KIND_BY_RAW;
}

export function isBoardCellStateRaw(value: unknown): value is BoardCellStateRaw {
  return typeof value === 'number' && value in BOARD_CELL_STATE_BY_RAW;
}

const NEXT_BOARD_CELL_STATE_BY_PREV_BOARD_CELL_STATE: Readonly<Record<BoardCellState, BoardCellState>> = {
  B: 'E',
  E: 'R',
  R: 'B',
};

const BALANCE_BY_BOARD_CELL_STATE: Readonly<Record<BoardCellState, number>> = {
  B: -1,
  E: 0,
  R: 1,
};

const RAW_BY_BOARD_CELL_KIND: Readonly<Record<BoardCellKind, BoardCellKindRaw>> = {
  fixed: 0,
  regular: 1,
};

const BOARD_CELL_KIND_BY_RAW: Readonly<Record<BoardCellKindRaw, BoardCellKind>> = {
  0: 'fixed',
  1: 'regular',
};

const RAW_BY_BOARD_CELL_STATE: Readonly<Record<BoardCellState, BoardCellStateRaw>> = {
  B: 0,
  E: 2,
  R: 1,
};

const BOARD_CELL_STATE_BY_RAW: Readonly<Record<BoardCellStateRaw, BoardCellState>> = {
  0: 'B',
  1: 'R',
  2: 'E',
};

export type BoardCellValue = readonly [kind: BoardCellKindRaw, state: BoardCellStateRaw];

export class BoardCell {
  get balance() {
    return BALANCE_BY_BOARD_CELL_STATE[this.#state];
  }

  get isEmpty() {
    return this.#state === 'E';
  }

  get isFilled() {
    return !this.isEmpty;
  }

  get isFixed() {
    return this.#kind === 'fixed';
  }

  get state() {
    return this.#state;
  }

  readonly #kind: BoardCellKind;
  readonly #state: BoardCellState;

  constructor(kind: BoardCellKind, state: BoardCellState) {
    this.#kind = kind;
    this.#state = state;
  }

  static create(state: BoardCellState) {
    return new this('regular', state);
  }

  static from(value: BoardCellValue) {
    return new this(BOARD_CELL_KIND_BY_RAW[value[0]], BOARD_CELL_STATE_BY_RAW[value[1]]);
  }

  equals(other: BoardCell) {
    return this.#state === other.#state;
  }

  next() {
    return this.isFixed ? this : new BoardCell(this.#kind, NEXT_BOARD_CELL_STATE_BY_PREV_BOARD_CELL_STATE[this.#state]);
  }

  toFixed() {
    return new BoardCell('fixed', this.#state);
  }

  valueOf(): BoardCellValue {
    return [RAW_BY_BOARD_CELL_KIND[this.#kind], RAW_BY_BOARD_CELL_STATE[this.#state]];
  }
}

export function isBoardCellValue(value: unknown): value is BoardCellValue {
  if (Array.isArray(value) && value.length === 2) {
    return isBoardCellKindRaw(value[0]) && isBoardCellStateRaw(value[1]);
  }

  return false;
}
