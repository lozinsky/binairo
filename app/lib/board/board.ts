import {
  DoubleMatrixRotation,
  type Matrix,
  type MatrixReversible,
  type MatrixRotatable,
  type MatrixSelection,
  QuadrupleMatrixRotation,
} from '~/lib/matrix';
import { assert } from '~/shared/assert';
import { expectToBeDefined } from '~/shared/expect';

import type { BoardCell, BoardCellState } from './board-cell';

import { BoardLine, type BoardLineValue, isBoardLineValue } from './board-line';

export type BoardOrientation = 'landscape' | 'portrait';

const NEXT_BOARD_ORIENTATION_BY_PREV_BOARD_ORIENTATION: Readonly<Record<BoardOrientation, BoardOrientation>> = {
  landscape: 'portrait',
  portrait: 'landscape',
};

export type BoardValue = readonly BoardLineValue[];

export class Board implements Iterable<BoardLine>, Matrix<BoardLine>, MatrixReversible<Board>, MatrixRotatable<Board> {
  get length() {
    return this.#lines.length;
  }

  get orientation() {
    return this.#orientation;
  }

  get progress() {
    return this.#lines.reduce((progress, line) => progress + line.progress, 0) / this.#lines.length;
  }

  readonly #lines: readonly BoardLine[];
  readonly #orientation: BoardOrientation;

  constructor(orientation: BoardOrientation, lines: readonly BoardLine[]) {
    for (const line of lines) {
      assert(line.length === lines.length, 'Expected line length to be equal to board size');
    }

    this.#lines = lines;
    this.#orientation = orientation;
  }

  static blank(size: number) {
    return new this(
      'portrait',
      Array.from({ length: size }, () => BoardLine.blank(size)),
    );
  }

  static create(states: ReadonlyArray<readonly BoardCellState[]>) {
    return new this(
      'portrait',
      states.map((states) => BoardLine.create(states)),
    );
  }

  static from(value: BoardValue) {
    return new this(
      'portrait',
      value.map((value) => BoardLine.from(value)),
    );
  }

  at(index: number) {
    return this.#lines.at(index);
  }

  entries() {
    return this.#lines.entries();
  }

  replace(index: number, callback: (line: BoardLine) => BoardLine) {
    return new Board(this.#orientation, this.#lines.with(index, callback(expectToBeDefined(this.#lines.at(index)))));
  }

  replaceBy(selection: MatrixSelection, callback: (cell: BoardCell) => BoardCell) {
    const content = selection.execute(this);
    const lines = this.#lines.map((line) => {
      const cells = Array.from(line);

      for (const cell of content) {
        const index = line.indexOf(cell);

        if (index === -1) {
          continue;
        }

        cells[index] = callback(expectToBeDefined(cells[index]));
      }

      return new BoardLine(cells);
    });

    return new Board(this.#orientation, lines);
  }

  reverse() {
    return new Board(this.#orientation, this.#lines.toReversed());
  }

  rotate() {
    const orientation = NEXT_BOARD_ORIENTATION_BY_PREV_BOARD_ORIENTATION[this.#orientation];

    if (this.#lines.length === 0) {
      return new Board(orientation, []);
    }

    const lines: BoardLine[] = [];

    for (const index of expectToBeDefined(this.#lines[0]).keys()) {
      const cells: BoardCell[] = [];

      for (const line of this.#lines) {
        cells.push(expectToBeDefined(line.at(index)));
      }

      lines.push(new BoardLine(cells));
    }

    return new Board(orientation, lines);
  }

  slice(start: number, end?: number) {
    return this.#lines.slice(start, end);
  }

  [Symbol.iterator]() {
    return this.#lines[Symbol.iterator]();
  }

  toDoubleRotation() {
    return DoubleMatrixRotation.run<Board>(this);
  }

  toQuadrupleRotation() {
    return QuadrupleMatrixRotation.run<Board>(this);
  }

  valueOf(): BoardValue {
    return this.#lines.map((line) => line.valueOf());
  }
}

export function isBoardValue(value: unknown): value is BoardValue {
  return Array.isArray(value) && value.every(isBoardLineValue);
}
