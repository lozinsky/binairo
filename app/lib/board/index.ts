export * from './board-cell';
export * from './board-line';
export * from './board';
