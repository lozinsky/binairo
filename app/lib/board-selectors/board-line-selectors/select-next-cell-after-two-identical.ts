import { type BoardCell, type BoardLine } from '~/lib/board';
import { expectToBeDefined } from '~/shared/expect';

export function selectNextCellAfterTwoIdentical(target: BoardLine) {
  for (const line of target.toDoubleRotation()) {
    for (const [start, head] of line.entries()) {
      if (head.isFilled) {
        const tail: BoardCell[] = [];

        for (const [end, cell] of line.entries()) {
          if (start >= end) {
            continue;
          }

          tail.push(cell);
        }

        if (tail.length >= 2) {
          const middle = expectToBeDefined(tail[0]);
          const next = expectToBeDefined(tail[1]);

          if (head.equals(middle) && next.isEmpty) {
            return { cell: next };
          }
        }
      }
    }
  }
}
