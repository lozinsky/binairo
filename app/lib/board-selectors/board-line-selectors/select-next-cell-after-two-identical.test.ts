import { expect, test } from 'vitest';

import { BoardCellState, BoardLine } from '~/lib/board';

import { selectNextCellAfterTwoIdentical } from './select-next-cell-after-two-identical';

test.each([
  BoardLine.create([BoardCellState.R, BoardCellState.R, BoardCellState.E, BoardCellState.E]),
  BoardLine.create([BoardCellState.E, BoardCellState.E, BoardCellState.R, BoardCellState.R]),
  BoardLine.create([BoardCellState.R, BoardCellState.E, BoardCellState.E, BoardCellState.E]),
  BoardLine.create([BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.R]),
])('selects next cell after two identical', (target) => {
  expect(selectNextCellAfterTwoIdentical(target)).toMatchSnapshot();
});
