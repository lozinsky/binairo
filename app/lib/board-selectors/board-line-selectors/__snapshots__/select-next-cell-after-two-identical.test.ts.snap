// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`selects next cell after two identical 1`] = `
{
  "cell": [⬜],
}
`;

exports[`selects next cell after two identical 2`] = `
{
  "cell": [⬜],
}
`;

exports[`selects next cell after two identical 3`] = `undefined`;

exports[`selects next cell after two identical 4`] = `undefined`;
